src/CSRMultiplication/SparseMatrixMultiplication.fs

```diff
@@ -0,0 +1,206 @@
+                    while (iPointer < csrRowPointersBuffer.[i + 1] && jPointer < cscColumnPointersBuffer.[j + 1]) do
+                        if csrColumnsBuffer.[iPointer] < cscRowsBuffer.[jPointer] then iPointer <- iPointer + 1
+                        elif csrColumnsBuffer.[iPointer] > cscRowsBuffer.[jPointer] then jPointer <- jPointer + 1
+                        else 
+                            localResultBuffer <- localResultBuffer +
+                                csrValuesBuffer.[iPointer] * cscValuesBuffer.[jPointer]
+                            iPointer <- iPointer + 1
+                            jPointer <- jPointer + 1   
+                    resultBuffer.[i * cscMatrixColumnCount + j] <- localResultBuffer 
+            @>
+
+        let (kernel, kernelPrepare, kernelRun) = provider.Compile command 
+        let ndRange = _2D(csrMatrixRowCount, cscMatrixColumnCount)
+        kernelPrepare 
+            ndRange 
+            resultMatrix 
+            csrMatrix.GetValues 
+            csrMatrix.GetColumns 
+            csrMatrix.GetRowPointers 
+            cscMatrix.GetValues 
+            cscMatrix.GetRows 
+            cscMatrix.GetColumnPointers
+        commandQueue.Add(kernelRun()).Finish() |> ignore
+        commandQueue.Add(resultMatrix.ToHost provider).Finish() |> ignore
+
+        (fun i j -> resultMatrix.[i * cscMatrixColumnCount + j])
+        |> Array2D.init csrMatrixRowCount cscMatrixColumnCount
+
+    let multiplySpMSpM2 
+        (provider: ComputeProvider) 
+        (commandQueue: CommandQueue) 
+        (csrMatrixRight: CSRMatrix.CSRMatrix) 
+        (csrMatrixLeft: CSRMatrix.CSRMatrix) = 
+
+        let leftMatrixRowCount = csrMatrixLeft |> CSRMatrix.rowCount
+        let leftMatrixColumnCount = csrMatrixLeft |> CSRMatrix.columnCount
+        let rightMatrixRowCount = csrMatrixRight |> CSRMatrix.rowCount
+        let rightMatrixColumnCount = csrMatrixRight |> CSRMatrix.columnCount
+        if leftMatrixColumnCount <> rightMatrixRowCount then  failwith "fail"
+
+        let resultMatrix = Array.zeroCreate<float> (leftMatrixRowCount * rightMatrixColumnCount)
+        let command = 
+            <@
+                fun (ndRange: _2D)
+                    (resultBuffer: float[]) 
+                    (leftCsrValuesBuffer: float[]) 
+                    (leftCsrColumnsBuffer: int[]) 
+                    (leftCsrRowPointersBuffer: int[]) 
+                    (rightCsrValuesBuffer: float[]) 
+                    (rightCsrColumnsBuffer: int[]) 
+                    (rightCsrRowPointersBuffer: int[])  ->
+
+                    let i = ndRange.GlobalID0
+                    let j = ndRange.GlobalID1
+                    for k in rightCsrRowPointersBuffer.[i] .. rightCsrRowPointersBuffer.[i + 1] - 1 do
+                        let mutable localResultBuffer = resultBuffer.[j * rightMatrixColumnCount + rightCsrColumnsBuffer.[k]]
```

Comment:
This may be a reason of the unstable behavior mentioned in this PR description. Read and write operations with ```resultBuffer``` is not synchronized between threads. Moreover, reading and writing are not atomic operations.